azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +64,152 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
+                    self._retrieve_certificates()
+
+                # Retrieve the other documents if we have either a Fabric or FastTrack change
+                if self.ext_conf.changed or retrieval_mode == GoalStateRetrievalMode.Full:
+                    self._retrieve_hosting_env()
+                    self._retrieve_shared_conf()
+                    self._retrieve_remote_access()
+
+    @staticmethod
+    def fetch_goal_state(wire_client, ext_config_retriever):
+        """
+        Fetches the goal state, not including any nested properties (such as remote access).
+        """
+        return GoalState(wire_client, ext_config_retriever)
 
-            uri = findtext(xml_doc, "SharedConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.shared_conf = SharedConfig(xml_text)
+    @staticmethod
+    def fetch_limited_goal_state(wire_client, ext_config_retriever):
+        """
+        Fetches only the wire server GoalState object. Does not include the extensions config
+        """
+        return GoalState(wire_client, ext_config_retriever, retrieval_mode=GoalStateRetrievalMode.Limited)
 
-            uri = findtext(xml_doc, "Certificates")
+    @staticmethod
+    def fetch_full_goal_state(wire_client, ext_config_retriever):
+        """
+        Fetches the full goal state, including nested properties (such as remote access).
+        """
+        return GoalState(wire_client, ext_config_retriever, retrieval_mode=GoalStateRetrievalMode.Full)
+
+    def _retrieve_hosting_env(self):
+        try:
+            uri = findtext(self._xml_doc, "HostingEnvironmentConfig")
             if uri is None:
-                self.certs = None
+                logger.warn("HostingEnvironmentConfig url doesn't exist in goal state")
             else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+                xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header())
+                self.hosting_env = HostingEnv(xml_text)
+        except Exception as e:
+            self._log_document_retrieval_failure("hosting environment", e)
+            raise
+
+    def _retrieve_certificates(self):
+        try:
+            uri = findtext(self._xml_doc, "Certificates")
+            if uri is not None:
+                xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header_for_cert())
                 self.certs = Certificates(xml_text)
+        except Exception as e:
+            self._log_document_retrieval_failure("certificates", e)
+            raise
 
-            uri = findtext(xml_doc, "ExtensionsConfig")
+    def _retrieve_shared_conf(self):
+        try:
+            uri = findtext(self._xml_doc, "SharedConfig")
             if uri is None:
-                self.ext_conf = ExtensionsConfig(None)
+                logger.warn("SharedConfig url doesn't exist in goal state")
```

Comment:
The previous code would assume the SharedConfig was always there and would have raised an error if it wasn't. What are the implications of changing this to a warning?

Reply:
same question for Hosting Environment

Reply:
Switched this to error. The goal was to be clearer about the issue in the logs.